include/exadg/convection_diffusion/preconditioners/multigrid_preconditioner.cpp

```diff
@@ -135,14 +146,25 @@ MultigridPreconditioner<dim, Number>::update()
       vector_multigrid_type_ptr  = &vector_multigrid_type_copy;
     }
 
-    set_velocity(*vector_multigrid_type_ptr);
-  }
+    unsigned int const fine_level   = this->get_number_of_levels() - 1;
+    unsigned int const coarse_level = 0;
 
-  // Once the operators are updated, the update of smoothers and the coarse grid solver is generic
-  // functionality implemented in the base class
+    // copy velocity to finest level
+    this->get_operator(fine_level)->set_velocity_copy(*vector_multigrid_type_ptr);
 
-  update_smoothers();
+    // interpolate velocity from fine to coarse level
+    for(unsigned int level = fine_level; level > coarse_level; --level)
```

Comment:
What about introducing `for_all_levels_reversed()`?

Reply:
The problem is that we have the condition `level > coarse_level`, because the functionality here is to "compare" two levels with each other. Writing `level-0` and `level-1` versus `level+1` and `level+0` is a convention. Hence, I do not consider this a classical "for-loop" over the levels. In this case, I consider the use of lambda error prone (and would agree here with the general comment by @kronbichler).

A generic function as suggested above would need to include an implementation of the index logic inside the function. Any suggestions regarding name and interface of such a generic function? `(transfer_)from_fine_to_all_coarser_levels([](uint, uint){})` with a lambda taking two parameters `unsigned int fine_level`/`unsigned int coarse_level` as arguments?

Reply:
You are definitely right that this is another type of functionality currently appearing many times in the code.

Reply:
see https://github.com/exadg/exadg/pull/456/commits/789280dc6864931adfa3470d5bd4d0256700ce1e